### PR TITLE
Handle base dir in get manifest hash

### PR DESCRIPTION
**Pull request by siu**
CLCACHE_BASEDIR is not used to cleanup the command line when computing the manifest hash. This is a problem because invocations of clcache from different basedirs will create different manifests for equivalent object files.

### CODE REVIEW

File: clcache.py [hunk 1]
```diff
@@ -232,6 +232,9 @@ def getManifestHash(compilerBinary, commandLine, sourceFile):
         # One of the few exceptions to this rule is the /MP switch, which only
         # defines how many compiler processes are running simultaneusly.
         commandLine = [arg for arg in commandLine if not arg.startswith("/MP")]
+        baseDir = os.environ.get('CLCACHE_BASEDIR')
+        if baseDir:
+            commandLine = [arg.replace(baseDir, BASEDIR_REPLACEMENT) for arg in commandLine]
```
Comment (frerich) [inline clcache.py]: You could reuse the existing `collapseBasedirToPlaceholder` here and use

```python
commandLine = [collapseBasedirToPlaceholder(arg) for arg in commandLine if not arg.startswith("/MP")]
```

in line 234.

Comment (siu) [inline clcache.py]: That wouldn't work with the current implementation of collapseBasedirToPlaceholder as it assumes that the BASEDIR appears in the first position of the argument and that the argument is in normcase.

Here arg can be of the form -IC:\basedir\... which doesn't satisfy any of those conditions. We shouldn't change the case either.

collapseBasedirToPlaceholder could be changed like this to make it work:

```
 def collapseBasedirToPlaceholder(path):
     baseDir = normalizeBaseDir(os.environ.get('CLCACHE_BASEDIR'))
-    if baseDir is None:
-        return path
-    else:
-        assert path == os.path.normcase(path)
-        assert baseDir == os.path.normcase(baseDir)
-        if path.startswith(baseDir):
-            return path.replace(baseDir, BASEDIR_REPLACEMENT, 1)
-        else:
-            return path
+    if baseDir:
+        replacement = re.compile(re.escape(baseDir), re.IGNORECASE)
+        return replacement.sub(BASEDIR_REPLACEMENT, path)
+    return path
```

Comment (frerich) [inline clcache.py]: Ah, true - as it is, `collapseBasedirToPlaceholder` is not applicable. While thinking about whether it makes sense to adjust it as you suggested, it occurred to me that maybe clcache shouldn't always blindly replace the basedir value.

For instance, should it really replace the basedir in preprocessor definitions passed via `-D` on the command line?

Comment (frerich) [inline clcache.py]: @siu -- ping? :)

Comment (siu) [inline clcache.py]: Yes, I don't think it is correct to apply the replacement blindly. I can instead add a list of parameters and apply the replacement only to those.

In that case, which arguments should be in the white list? I had a look here (https://msdn.microsoft.com/en-us/library/fwkeyyhe.aspx), these are the arguments that seem good candidates:

 - "/I" https://msdn.microsoft.com/en-us/library/73f9s62w.aspx
 - "/AI" https://msdn.microsoft.com/en-us/library/x1x72k9t.aspx
 - "/FU" https://msdn.microsoft.com/en-us/library/81ex1b0a.aspx
 - "/Yu" https://msdn.microsoft.com/en-us/library/z0atkd6c.aspx

Comment (frerich) [inline clcache.py]: Indeed, the first three sound like good candidates. I suggest the replacement should be applied to their argument -- we can extend (or shorten) that list later as we learn about new cases.

`/Yu` doesn't seem important right now since clcache doesn't support invocations making use of precompiled headers anyway.

**Comment by siu** [comment]
Summary of the changes in this PR:
 * Added integration tests to show how clcache should behave with /I and /D arguments that include CLCACHE_BASEDIR
 * Apply collapseBasedirToPlaceholder to arguments that begin with /I, /AI, /FU
 * Change collapseBasedirToPlaceholder to find the basedir at any position and ignore the capitalization
 * Change normalizeBaseDir to remove trailing "/" and adapt unit tests. I am not sure about this one, will this break anything? Should I increase the MANIFEST_FILE_FORMAT_VERSION again?
 * Cleanup in TestBasedir
 * Make ASSETS_DIR in integration tests an absolute path (to simplify TestBasedir)

I also noticed that the integration tests have a lot of duplicated code, this could be simplified if all of them inherited from a common base class. Anyway that is for another PR.

Note: github shows the commit in a different order than they were applied.

File: clcache.py [hunk 2]
```diff
@@ -97,8 +97,8 @@ def childDirectories(path, absolute=True):
 def normalizeBaseDir(baseDir):
     if baseDir:
         baseDir = os.path.normcase(baseDir)
-        if not baseDir.endswith(os.path.sep):
-            baseDir += os.path.sep
+        if baseDir.endswith(os.path.sep):
+            baseDir = baseDir[0:-1]
```
Comment (frerich) [inline clcache.py]: Why did you need to change this?

File: clcache.py [hunk 3]
```diff
@@ -781,15 +788,10 @@ def expandBasedirPlaceholder(path):
...
-        assert path == os.path.normcase(path)
-        assert baseDir == os.path.normcase(baseDir)
-        if path.startswith(baseDir):
-            return path.replace(baseDir, BASEDIR_REPLACEMENT, 1)
-        else:
-            return path
+    if baseDir:
+        replacement = re.compile(re.escape(baseDir), re.IGNORECASE)
+        return replacement.sub(BASEDIR_REPLACEMENT, path)
+    return path
```
Comment (frerich) [inline clcache.py]: Any reason you dropped the `assert` statements? They seem to express that the function expects that both the given `path` as well as the `baseDir` are normalized, i.e. have not only the same upper/lowercase but also forward slashes converted to backslashes.

File: integrationtests.py [hunk 4]
```diff
@@ -24,7 +24,7 @@
 
 PYTHON_BINARY = sys.executable
 CLCACHE_SCRIPT = os.path.join(os.path.dirname(os.path.realpath(__file__)), "clcache.py")
-ASSETS_DIR = os.path.join("tests", "integrationtests")
+ASSETS_DIR = os.path.join(os.path.dirname(os.path.realpath(__file__)), "tests", "integrationtests")
```
Comment (frerich) [inline integrationtests.py]: I have a suspicion, but I'm not quite sure - why did you change this?

**Comment by siu** [comment]
That is true, some changes look pretty random without an explanation. I will try to explain myself better. I can reply you in the PR this time, but just to be clear, in the future do you prefer to have the explanation in each of the commits?

File: clcache.py (hunk 2 above)
Comment (siu) [inline clcache.py]: This was needed to make the integration test `TestBasedir.testBasedirIncludeArg` pass. There the include path is generated by `os.getcwd()` which doesn't include the trailing separator. As `normalizedBasedir` is used in `collapseBasedirToPlaceholder` the version with the trailing slash would not match.

**Comment by frerich** [comment]
Thanks - yes, it's very much appreciated if the commits themselves explain the reasoning behind the change. This is very useful when debugging things later on, where you typically check the history to see why some changes were done -- and it might be in a time or place where GitHub is not even available, so all the things discussed here are inaccessible.

I very much recommend [this page](http://chris.beams.io/posts/git-commit/) for some good guidelines on how to write commit logs, all future contributors (and your future self :-)) will be grateful! :-)

For what it's worth, it's very easy to update your commits with logs. With the command line client, you can run `git rebase --interactive <commit_on_which_you_based_your_branch>` and then change the action for each commit to perform (the word in the first column) to `reword`. Git will then replay all your commits, but on each commit it will pause and open an editor which allows you to edit the commit log.

File: clcache.py (hunk 3 above)
Comment (siu) [inline clcache.py]: Indeed I applied this change as a quick test and didn't give it too much thought. With the proposed implementation the `assert path == os.path.normcase(path)` does not hold. I can add back the other assert `assert baseDir == os.path.normcase(baseDir)`

File: integrationtests.py (hunk 4 above)
Comment (siu) [inline integrationtests.py]: This was changed only to simplify the code in TestBasedir. The cwd is changed with `cd()` and the relative path to the assets wouldn't work. If I rearrange the code there a bit this change is not needed.

**Comment by siu** [comment]
Thanks for the clarification. I will amend the commits once the review is done.

File: clcache.py (hunk 2 above)
Comment (frerich) [inline clcache.py]: I first wanted to insist on rather adjusting the test such that it appends a path separator (if needed) to the `os.getcwd()` return value instead of adjusting this function, but I guess if `os.getcwd()` does not return a trailing slash then that's a good indicator that we shouldn't do so either. So I'm okay with this. 👍

File: clcache.py (hunk 3 above)
Comment (frerich) [inline clcache.py]: Why would the `assert path == os.path.normcase(path)` not hold anymore? Ah, I guess it's because this function is not given paths anymore, but command line arguments, e.g. `/IC:\Foo\Bar`. That suggests that either the argument should be renamed (`path` does not fit anymore), or maybe the newly introduced `collapseBasedirInCmd` should not pass arguments to `collapseBasedirToPlaceholder` but rather the actual paths.

To extract the paths from the arguments, the code in `CommandLineAnalyzer.parseArgumentsAndFiles` could maybe get refactored to be more reusable (it already solves the issue of getting the values for command line arguments).

I guess for now, simply renaming the `path` argument to e.g. `argument` or maybe even just `s` would be fine, at least it's not misleading anymore.

File: integrationtests.py (hunk 4 above)
Comment (frerich) [inline integrationtests.py]: I see, makes sense! 👍

File: clcache.py (hunk 2 above)
Comment (frerich) [inline clcache.py]: I first wanted to insist on rather adjusting the test such that it appends a path separator (if needed) to the os.getcwd() return value instead of adjusting this function, but I guess if os.getcwd() does not return a trailing slash then that's a good indicator that we shouldn't do so either. So I'm okay with this. 👍

**Comment by siu** [comment]
@frerich I rewrote the PR with your suggestions.

 * Thanks for pointing to CommandLineAnalyzer.parseArgumentsAndFiles, it can be used from getManifestHash as it is and makes the changes in collapseBasedirToPlaceholder unnecessary.
 * The trailing separator still must be removed from normalizeBaseDir as the include arguments may not have it and it should match.
 * I kept the absolute ASSETS_DIR as it makes writing tests easier.

File: clcache.py [hunk 5]
```diff
@@ -232,9 +232,21 @@ def getManifestHash(compilerBinary, commandLine, sourceFile):
         # One of the few exceptions to this rule is the /MP switch, which only
         # defines how many compiler processes are running simultaneusly.
         commandLine = [arg for arg in commandLine if not arg.startswith("/MP")]
+        arguments, inputFiles = CommandLineAnalyzer.parseArgumentsAndInputFiles(commandLine)
+        collapseBasedirInCmdPath = lambda path: collapseBasedirToPlaceholder(os.path.normcase(os.path.abspath(path)))
+
+        commandLineArgs = []
+        projectSpecificArgs = ("AI", "I", "FU")
+        for k in sorted(arguments.keys()):
```
Comment (frerich) [inline clcache.py]: Why do you use `sorted` here?

Comment (siu) [inline clcache.py]: Actually the sorted is very important to have repeatable manifest hashes. As dictionaries in python are unordered the only way to guarantee that we generate the same hash every invocation is to iterate over the keys in the same order. It is the same case that #235 solved.

Comment (frerich) [inline clcache.py]: Ah, so this basically improves the manifest cache hit rate, i.e. it's an improvement but not directly related to the BASEDIR feature? Ok.

Comment (siu) [inline clcache.py]: It is going to have that side effect but at same time it is needed for correctness since the implementation changed to use a dictionary.

**Comment by frerich** [comment]
Thanks for being so patient and explaining all the changes to me, I don't see any reason not to merge this nice improvement - thanks a lot! :-)

**Comment by siu** [comment]
I am happy to iterate and give explanations if things are not clear, the end result is better ;) Thanks for merging.
